shared/lib_irradproc.h

```diff
@@ -628,9 +628,68 @@ double sun_rise_and_set(double* m_rts, double* h_rts, double* delta_prime, doubl
 * \param[out] needed_values[7] zenith topocentric zenith angle (degrees)
 * \param[out] needed_values[8] azimuth topocentric azimuth angle (degrees)
 */
+
+
 void calculate_spa(double jd, double lat, double lng, double alt, double pressure, double temp,
     double delta_t, double tilt, double azm_rotation, double ascension_and_declination[2], double needed_values[9]);
 
+/**
+ * Table for storing the recently computed solarpos_spa intermediate outputs
+ * The algorithm reuses the outputs from the last 3 days or so, so the hash table is emptied every 3 days to reduce size
+ * Latitude, Longitude, Altitude, Tilt and Azimuth are not in the key because they remain constant throughout
+ */
+struct spa_table_key {
+    double jd;
+    double delta_t;
+    int pressure;
+    int temp;
+    // these are both inputs and outputs (e.g. also stored in the output vector)
+    double ascension;
+    double declination;
+
+    bool operator==(const spa_table_key &other) const
+        { return (jd == other.jd
+                && delta_t == other.delta_t
+                && pressure == other.pressure
+                && temp == other.temp
+                && ascension == other.ascension
+                && declination == other.declination
+                );
+        }
+
+    spa_table_key(double j, double dt, double p, double t, double a, double d):
+        jd(j), delta_t(dt), ascension(a), declination(d)
+    {
+        int pressure_bucket = 10;
+        pressure = ((int)(p + pressure_bucket/2) / pressure_bucket) * pressure_bucket;
+        pressure = (int)pressure;
+        int temp_bucket = 5;
+        temp = ((int)(t + temp_bucket / 2) / temp_bucket) * temp_bucket;
+        temp = (int)temp;
+    }
+};
+
+template <>
+struct std::hash<spa_table_key>
+{
+  std::size_t operator()(const spa_table_key& k) const
+  {
+    using std::hash;
+    // Compute individual hash values for first, second, etc and combine them using XOR and bit shifting:
+    return 
+    ((((
+            ((((hash<double>()(k.jd)
```

Comment:
Have you tried hash_combine in C++17 for more robustness (i.e. less collisions)?

![image](https://github.com/user-attachments/assets/d1445283-2976-48d6-9555-aa51a22b1f3b)

Reply:
Once we've upgraded to C++17, this would be great!